File: py/tests/test_join.py
"""A safe left join for EAVS frames.

Mirrors r/tests/testthat/test-join.R. Fixtures model the two hazards by hand: a
Wisconsin town/village pair sharing one serial code (the fan-out) and a Maine
statewide row absent from y (the drop). Network-free.
"""

from __future__ import annotations

import warnings

import pandas as pd
import pytest

import tidyeavs


def join_panel() -> pd.DataFrame:
    return pd.DataFrame(
        {
            "year": [2020] * 4,
            "fips_code": ["82575", "82575", "55001000", "23"],
            "jurisdiction_name": [
                "TOWN OF ARBOR",
                "VILLAGE OF ARBOR",
                "COUNTY A",
                "MAINE UOCAVA",
            ],
            "state_abbr": ["WI", "WI", "WI", "ME"],
            "partic_total": [100.0, 50.0, 200.0, 7.0],
        }
    )


def join_jurisdictions() -> pd.DataFrame:
    return pd.DataFrame(
        {
            "year": [2020] * 3,
            "fips_code": ["82575", "82575", "55001000"],
            "jurisdiction_name": ["TOWN OF ARBOR", "VILLAGE OF ARBOR", "COUNTY A"],
            "state_abbr": ["WI", "WI", "WI"],
            "county_fips": [None, None, "55001"],
            "type": ["municipality", "municipality", "county"],
        }
    )


def test_shared_code_stops_the_join():
    with pytest.raises(ValueError, match="fan out"):
        tidyeavs.join(join_panel(), join_jurisdictions())


def test_fanout_message_points_at_jurisdiction_name():
    with pytest.raises(ValueError, match="jurisdiction_name"):
        tidyeavs.join(join_panel(), join_jurisdictions())


def test_disambiguating_resolves_the_pair_one_to_one():
    out = tidyeavs.join(
        join_panel(),
        join_jurisdictions(),
        by=["year", "fips_code", "jurisdiction_name"],
        unmatched="ignore",
    )
    assert len(out) == 4  # no fan-out
    town = out[out["jurisdiction_name"] == "TOWN OF ARBOR"].iloc[0]
    assert pd.isna(town["county_fips"])
    county = out[out["jurisdiction_name"] == "COUNTY A"].iloc[0]
    assert county["county_fips"] == "55001"


def test_multiple_all_keeps_the_expansion():
    out = tidyeavs.join(
        join_panel(),
        join_jurisdictions(),
        by=["year", "fips_code"],
        multiple="all",
        unmatched="ignore",
    )
    # 2 panel rows x 2 y rows for 82575 = 4, plus the county (1) and Maine (1).
    assert len(out) == 6


def test_unmatched_row_is_reported_but_kept():
    by = ["year", "fips_code", "jurisdiction_name"]
    with pytest.warns(UserWarning, match="matched nothing"):
        tidyeavs.join(join_panel(), join_jurisdictions(), by=by)
    with pytest.raises(ValueError, match="matched nothing"):
        tidyeavs.join(join_panel(), join_jurisdictions(), by=by, unmatched="error")
    out = tidyeavs.join(join_panel(), join_jurisdictions(), by=by, unmatched="ignore")
    maine = out[out["jurisdiction_name"] == "MAINE UOCAVA"]
    assert len(maine) == 1
    assert pd.isna(maine.iloc[0]["county_fips"])


def test_ignore_is_silent_and_inform_names_the_state():
    by = ["year", "fips_code", "jurisdiction_name"]
    with warnings.catch_warnings():
        warnings.simplefilter("error")  # any warning becomes an error
        tidyeavs.join(join_panel(), join_jurisdictions(), by=by, unmatched="ignore")
    with pytest.warns(UserWarning, match="ME"):
        tidyeavs.join(join_panel(), join_jurisdictions(), by=by)


def test_only_new_columns_are_added():
    out = tidyeavs.join(
        join_panel(),
        join_jurisdictions(),
        by=["year", "fips_code", "jurisdiction_name"],
        unmatched="ignore",
    )
    assert not any(c.endswith("_x") or c.endswith("_y") for c in out.columns)
    assert {"county_fips", "type"}.issubset(out.columns)
    assert list(out.columns).count("state_abbr") == 1  # kept from x, not doubled


def test_by_defaults_to_shared_year_and_fips_code():
    x = pd.DataFrame({"year": [2020], "fips_code": ["55001000"], "v": [1]})
    y = pd.DataFrame({"year": [2020], "fips_code": ["55001000"], "w": [2]})
    out = tidyeavs.join(x, y)
    assert out.iloc[0]["w"] == 2


def test_no_shared_key_columns_is_an_error():
    with pytest.raises(ValueError, match="No columns to join on"):
        tidyeavs.join(pd.DataFrame({"a": [1]}), pd.DataFrame({"b": [2]}))


def test_by_naming_a_missing_column_is_an_error():
    x = pd.DataFrame({"year": [2020], "fips_code": ["1"]})
    y = pd.DataFrame({"year": [2020], "fips_code": ["1"]})
    with pytest.raises(ValueError, match="not in both frames"):
        tidyeavs.join(x, y, by=["year", "nope"])

File: r/tests/testthat/test-join.R
## Fixtures model the two hazards by hand: a Wisconsin town/village pair sharing
## one serial code (the fan-out) and a Maine statewide row absent from y (the
## drop). Small enough that every row count is checkable.

join_panel <- function() {
  tibble::tibble(
    year = rep(2020L, 4),
    fips_code = c("82575", "82575", "55001000", "23"),
    jurisdiction_name = c("TOWN OF ARBOR", "VILLAGE OF ARBOR",
                          "COUNTY A", "MAINE UOCAVA"),
    state_abbr = c("WI", "WI", "WI", "ME"),
    partic_total = c(100, 50, 200, 7)
  )
}

join_jurisdictions <- function() {
  tibble::tibble(
    year = rep(2020L, 3),
    fips_code = c("82575", "82575", "55001000"),
    jurisdiction_name = c("TOWN OF ARBOR", "VILLAGE OF ARBOR", "COUNTY A"),
    state_abbr = c("WI", "WI", "WI"),
    county_fips = c(NA, NA, "55001"),
    type = c("municipality", "municipality", "county")
  )
}

test_that("a shared code stops the join instead of fanning it out", {
  expect_error(
    eavs_join(join_panel(), join_jurisdictions()),
    "fan out"
  )
})

test_that("the fan-out message points at jurisdiction_name when it can", {
  expect_error(
    eavs_join(join_panel(), join_jurisdictions()),
    "jurisdiction_name"
  )
})

test_that("adding jurisdiction_name to by resolves the pair one-to-one", {
  out <- eavs_join(
    join_panel(), join_jurisdictions(),
    by = c("year", "fips_code", "jurisdiction_name"),
    unmatched = "ignore"
  )
  expect_equal(nrow(out), 4L)             # no fan-out
  town <- out[out$jurisdiction_name == "TOWN OF ARBOR", ]
  expect_true(is.na(town$county_fips))
  county <- out[out$jurisdiction_name == "COUNTY A", ]
  expect_equal(county$county_fips, "55001")
})

test_that("multiple = 'all' keeps the expansion", {
  out <- eavs_join(
    join_panel(), join_jurisdictions(),
    by = c("year", "fips_code"),
    multiple = "all", unmatched = "ignore"
  )
  # 2 panel rows x 2 y rows for 82575 = 4, plus the county (1) and Maine (1).
  expect_equal(nrow(out), 6L)
})

test_that("an unmatched row is reported but kept with NA", {
  by <- c("year", "fips_code", "jurisdiction_name")
  expect_message(
    eavs_join(join_panel(), join_jurisdictions(), by = by),
    "matched nothing"
  )
  expect_error(
    eavs_join(join_panel(), join_jurisdictions(), by = by, unmatched = "error"),
    "matched nothing"
  )
  out <- eavs_join(join_panel(), join_jurisdictions(), by = by, unmatched = "ignore")
  maine <- out[out$jurisdiction_name == "MAINE UOCAVA", ]
  expect_equal(nrow(maine), 1L)
  expect_true(is.na(maine$county_fips))
})

test_that("unmatched = 'ignore' is silent and unmatched = 'inform' names the state", {
  by <- c("year", "fips_code", "jurisdiction_name")
  expect_silent(
    eavs_join(join_panel(), join_jurisdictions(), by = by, unmatched = "ignore")
  )
  expect_message(
    eavs_join(join_panel(), join_jurisdictions(), by = by),
    "ME"
  )
})

test_that("only y's new columns are added, with no .x/.y suffixes", {
  out <- eavs_join(
    join_panel(), join_jurisdictions(),
    by = c("year", "fips_code", "jurisdiction_name"),
    unmatched = "ignore"
  )
  expect_false(any(grepl("\\.x$|\\.y$", names(out))))
  expect_true(all(c("county_fips", "type") %in% names(out)))
  expect_equal(sum(names(out) == "state_abbr"), 1L)   # kept from x, not doubled
})

test_that("by defaults to the shared year and fips_code", {
  x <- tibble::tibble(year = 2020L, fips_code = "55001000", v = 1)
  y <- tibble::tibble(year = 2020L, fips_code = "55001000", w = 2)
  out <- eavs_join(x, y)
  expect_equal(out$w, 2)
})

test_that("no shared key columns is an error", {
  expect_error(
    eavs_join(tibble::tibble(a = 1), tibble::tibble(b = 2)),
    "No columns to join on"
  )
})

test_that("by naming a column missing from a frame is an error", {
  x <- tibble::tibble(year = 2020L, fips_code = "1")
  y <- tibble::tibble(year = 2020L, fips_code = "1")
  expect_error(eavs_join(x, y, by = c("year", "nope")), "not in both frames")
})
